### PR TITLE
ELL GPU kernels

### LINKED ISSUE #3: ELLPACK GPU kernels
- [ ] requires #2 
- [ ] provide GPU implementation of the kernels in `gpu/matrix/ell_kernels.cu`
- [ ] compare GPU implementation with reference kernels in `gpu/test/matrix/ell_kernels.cpp`

### PR DESCRIPTION
**Pull request by yanjen**
This closes #3 
Should we do multiple right-hand side spmv now?

Also, are dense matrices row major or column major on GPU?
When we allocate multiple b vectors, they should be in row major or column major?

### CODE REVIEW

**Comment by yanjen** [comment]
Oohs, I should have waited for replace padding with stride to close.

**Comment by yanjen** [comment]
I have updated the files with the latest updates in ginkgo.

**Comment by gflegar** [comment]
Could you again resolve the conflicts? We had to do some modifications to the commit history after the last PR.

Btw, do you have clang-format enabled? And did you do any changes to the `.clang-format` script, because it complained some files were not formatted properly when I was merging the last PR.

**Comment by gflegar** [comment]
Let's also hope that resolving the conflicts again will remove all the strange stuff I see in the changes - there it seems that we are once again replacing padding with stride, even though the develop branch doesn't contain any references to padding.

**Comment by yanjen** [comment]
Dear @gflegar you are right, my clang format was not working.
I have resolved the conflicts again.
Hope this works. 😄

**Comment by gflegar** [comment]
Good, the change list looks better now, I'll take a look.

File: gpu/test/matrix/ell_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+
+#include <gtest/gtest.h>
+
+
+#include <core/base/exception.hpp>
+#include <core/base/executor.hpp>
+#include <core/matrix/dense.hpp>
+#include <core/test/utils.hpp>
+#include "core/base/exception_helpers.hpp"
```
Comment (gflegar) [inline gpu/test/matrix/ell_kernels.cpp]: nit: `#include <core/base/exception_helpers.hpp>`

File: gpu/test/matrix/ell_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        dy = Vec::create(gpu);
+        dy->copy_from(y.get());
+        dalpha = Vec::create(gpu);
+        dalpha->copy_from(alpha.get());
+        dbeta = Vec::create(gpu);
+        dbeta->copy_from(beta.get());
+    }
+
+
+    void set_up_apply_data_with_stride(int stride, int max_nonzeros_per_row)
```
Comment (gflegar) [inline gpu/test/matrix/ell_kernels.cpp]: Do we need two functions here?

Isn't it enough to use default parameters like this:

```c++
void set_up_apply_data(int stride = 0, int max_nonzeros_per_row = 0)
```

and we get both behaviors with a single function. Or am I missing something?

File: gpu/test/matrix/ell_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+TEST_F(Ell, SimpleApplyIsEquivalentToRef)
+{
+    set_up_apply_data();
+
+    mtx->apply(y.get(), expected.get());
+    dmtx->apply(dy.get(), dresult.get());
+
+    auto result = Vec::create(ref);
+    result->copy_from(dresult.get());
```
Comment (gflegar) [inline gpu/test/matrix/ell_kernels.cpp]: no need to explicitly copy back to host anymore - `ASSERT_MTX_NEAR` will do that for you.

**Comment by gflegar** [comment]
> Also, are dense matrices row major or column major on GPU?
> When we allocate multiple b vectors, they should be in row major or column major?

Dense matrix is always row major, the storage format doesn't depend on the executor where it's run.
Since we represent a set of vectors as Dense matrix, this is also always row major.

> Should we do multiple right-hand side spmv now?

You mean in this PR? Not sure, haven't yet though if there are some interesting optimizations that can be done for multiple RHS. @hartwiganzt have you? In any case CSR also doesn't yet support this, so we can postpone it for later.

File: gpu/matrix/ell_kernels.cu [hunk 4]
```diff
@@ -46,22 +47,117 @@ namespace gpu {
...
 
 
+constexpr int default_block_size = 512;
+
+
+namespace {
+
+
+template <typename ValueType, typename IndexType>
+// clang-format off
```
[the remark this answers is not in the document]
Comment (yanjen) [reply gpu/matrix/ell_kernels.cu]: OK, I have fixed that.

File: gpu/test/matrix/ell_kernels.cpp (hunk 2 above)
Comment (yanjen) [reply gpu/test/matrix/ell_kernels.cpp]: I have combined the two functions into one.

**Comment by hartwiganzt** [comment]
Multi-Vector SpMV is on the list, but not urgent. We first need to have the Krylov solvers prepared for that before it makes any sense to have multi-vector SpMV.

File: gpu/matrix/ell_kernels.cu [hunk 5]
```diff
@@ -46,22 +47,99 @@ namespace gpu {
...
 
 
+namespace {
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void advanced_spmv_kernel(
+    size_type num_rows, const ValueType *__restrict__ alpha,
+    const ValueType *__restrict__ val, const IndexType *__restrict__ col,
+    const size_type stride, const size_type max_nonzeros_per_row,
```
Comment (gflegar) [inline gpu/matrix/ell_kernels.cu]: also remove `const` here

File: gpu/test/matrix/ell_kernels.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    std::unique_ptr<Vec> gen_mtx(int num_rows, int num_cols, int min_nnz_row)
+    {
+        return gko::test::generate_random_matrix<Vec>(
+            ref, num_rows, num_cols,
+            std::uniform_int_distribution<>(min_nnz_row, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine);
+    }
+
+    void set_up_apply_data(int stride = 0, int max_nonzeros_per_row = 0)
```
Comment (gflegar) [inline gpu/test/matrix/ell_kernels.cpp]: are you sure this works fine? I don't see `stride` or `max_nonzeros_per_row` being used anywhere in this function...

File: gpu/matrix/ell_kernels.cu (hunk 5 above)
Comment (yhmtsai) [reply gpu/matrix/ell_kernels.cu]: fix it.

File: gpu/test/matrix/ell_kernels.cpp (hunk 6 above)
Comment (yhmtsai) [reply gpu/test/matrix/ell_kernels.cpp]: @gflegar 
Fix it.

I have a problem with the ordering of parameters.
The matrix of nonzeros size is `stride X max_nonzeros_per_row`.
We initial ELL format like `Mtx::create(ref, 0, 0, max_nonzeros_per_row, stride);`
Is it better we change the ordering of stride and max_nonzeros_per_row?

Comment (gflegar) [reply gpu/test/matrix/ell_kernels.cpp]: Actually, I would have `max_nonzeros_per_row` before the `stride` in the parameter list. The reason for that is that the user is more likely to change the default `max_nonzeros_per_row` than the `stride` parameter:

1.  `Mtx::create(ref, m, n)` creates an `m`-by-`n` matrix, with 0 nonzeros, and stride set to `m`.
2.  `Mtx::create(ref, m, n, nnz)` creates an `m`-by-`n` matrix with `nnz` nonzeros, and stride set to `m`.
3.  `Mtx::create(ref, m, n, nnz, stride)` creates an `m`-by-`n` matrix with `nnz` nonzeros, and stride set to `stride`.

If we were to replace the order of parameters, you lose the possibility to specify option 2. easily, and I feel this is the most useful one, definitely more useful than `Mtx::create(ref, m, n, stride)` (a matrix with 0 nonzeros and stride set to `stride`).

**Pull request merged**